README.md
# gf-worldpay

[![Packagist Version](https://img.shields.io/packagist/v/itinerisltd/gf-worldpay.svg)](https://packagist.org/packages/itinerisltd/gf-worldpay)
[![PHP from Packagist](https://img.shields.io/packagist/php-v/itinerisltd/gf-worldpay.svg)](https://packagist.org/packages/itinerisltd/gf-worldpay)
[![Packagist Downloads](https://img.shields.io/packagist/dt/itinerisltd/gf-worldpay.svg)](https://packagist.org/packages/itinerisltd/gf-worldpay)
[![GitHub License](https://img.shields.io/github/license/itinerisltd/gf-worldpay.svg)](https://github.com/ItinerisLtd/gf-worldpay/blob/master/LICENSE)
[![Hire Itineris](https://img.shields.io/badge/Hire-Itineris-ff69b4.svg)](https://www.itineris.co.uk/contact/)


Gravity forms add-on for WorldPay.

<!-- START doctoc generated TOC please keep comment here to allow auto update -->
<!-- DON'T EDIT THIS SECTION, INSTEAD RE-RUN doctoc TO UPDATE -->


- [Minimum Requirements](#minimum-requirements)
- [Installation](#installation)
- [Setup](#setup)
- [Security Concerns about WorldPay HTML API](#security-concerns-about-worldpay-html-api)
- [Not Issue](#not-issue)
- [Features](#features)
- [Not Supported / Not Implemented](#not-supported--not-implemented)
- [Best Practices](#best-practices)
  - [HTTPS Everywhere](#https-everywhere)
  - [Payment Status](#payment-status)
- [Test Sandbox](#test-sandbox)
- [FAQ](#faq)
  - [GF WorldPay is Missing on Form Settings](#gf-worldpay-is-missing-on-form-settings)
- [Public API](#public-api)
  - [Build URL for continuing confirmation](#build-url-for-continuing-confirmation)
  - [Redirect URL Retrieval Failure Handling](#redirect-url-retrieval-failure-handling)
- [Preflight](#preflight)
- [Coding](#coding)
  - [Required Reading List](#required-reading-list)
  - [Gravity Forms](#gravity-forms)
- [Author Information](#author-information)
- [Feedback](#feedback)
- [Change log](#change-log)
- [License](#license)

<!-- END doctoc generated TOC please keep comment here to allow auto update -->

## Minimum Requirements

- PHP v7.2
- php-curl
- WordPress v4.9.8
- Gravity Forms v2.3.3.2

## Installation

```bash
$ composer require itinerisltd/gf-worldpay
```

## Setup

[Payment response(redirection)](http://support.worldpay.com/support/kb/bg/htmlredirect/htmlredirect.htm#rhtml/Telling_your_shopper_about.htm#_Payment_Response_messages) and [Enhancing security with MD5](http://support.worldpay.com/support/kb/bg/htmlredirect/htmlredirect.htm#rhtml/Enhancing_security_with_MD5.htm%3FTocPath%3D_____10) are mandatory.

In the Integration Setup for your installation using [the Merchant Interface > Installations option](http://support.worldpay.com/support/kb/bg/customisingadvanced/custa6011.html):

1. Enable **Enable the Shopper Response**
1. Enable **Payment Response enabled?**
1. Enter `<wpdisplay item=MC_callback>` as **Payment Response URL**
1. Enter a 25-char random passphrase as **Payment Response password**
1. Enter a 30-char random passphrase as **MD5 secret for transactions**
1. Enter `instId:amount:currency:cartId` as **SignatureFields**

Note that WorldPay truncate long **Payment Response password** without notices!

## Security Concerns about WorldPay HTML API

- Leaking **MD5 secret for transactions**
  * Allow evil hackers to set up fake checkout pages, pretending to be the merchant
  * WorldPay would accept these checkouts and charges the credit cards
  * Money goes into the merchant's account
- Leaking **Payment Response password**
  * Allow evil hackers to pretending to be WorldPay
  * WordPress would accept evil hackers' payment callbacks and changes entries' payment statuses

## Not Issue

If **Payment Response password**(also known as`callbackPW`) is incorrect, `InvalidResponseException` is throw to *stop the world*.
Credit card holders see white screen of death or stuck in "wait for redirection" page in such case.

## Features

- [Enhancing security with MD5](http://support.worldpay.com/support/kb/bg/htmlredirect/htmlredirect.htm#rhtml/Enhancing_security_with_MD5.htm%3FTocPath%3D_____10)
- [Gravity Forms Logging](https://docs.gravityforms.com/logging-and-debugging/)
- [Gravity Forms Notification Events](https://docs.gravityforms.com/gravity-forms-notification-events/)
- [Gravity Forms Confirmation](https://docs.gravityforms.com/configuring-confirmations-in-gravity-forms/)
- [Gravity Forms Conditional Logic](https://docs.gravityforms.com/enable-conditional-logic/)

## Not Supported / Not Implemented

- Shipping address
- Reject according to fraud check results
- Token payment
- Recurring payment
- Refund
- Void

## Best Practices

### HTTPS Everywhere

Although WorldPay accepts insecure HTTP sites, you should **always use HTTPS** to protect all communication.

### Payment Status

Always double check payment status on WorldPay Merchant Interface.

## Test Sandbox

Use this [test credit card](http://support.worldpay.com/support/kb/bg/pdf/181450-test-transaction-f.pdf).

## FAQ

### GF WorldPay is Missing on Form Settings

Gravity Forms capabilities behave differently on multi-user sites and its documents are incomplete.
If GF WorldPay is missing on form settings, grant yourself `gf_worldpay` and `gf_worldpay_uninstall` capabilities.
See: [https://docs.gravityforms.com/role-management-guide/](https://docs.gravityforms.com/role-management-guide/)

## Public API

### Build URL for continuing confirmation

`ConfirmationHandler::buildUrlFor(Entry $entry, int $ttlInSeconds = 3600): string`

Usage:
```php
<?php
$entryId = 123;
$rawEntry = GFAPI::get_entry($entryId);
if (is_wp_error($rawEntry)) {
    wp_die('Entry not found');
}

$url = ConfirmationHandler::buildUrlFor(
    new Entry($rawEntry),
    86400 // expires in 24 hours (24*3600=86400)
);

echo $url;
// https://example.com?entry=123&gf-worldpay-token=XXXXXXXXXXXX
```

Use Case:
With ["using confirmation query strings to populate a form based on another submission"](https://docs.gravityforms.com/using-confirmation-query-strings-to-populate-a-form-based-on-another-submission/):
1. User fills in formA
1. User completes WorldPay checkout form
1. User comes back and hits `CallbackHandler`
1. `CallbackHandler` sends user to formB according to confirmation settings
1. User arrives formB url with merged query strings

If the user quits before completing formB, you could use `ConfirmationHandler::buildUrlFor` generate a single-use, short-lived url for the user to resume formB.

Note:
- The url continues Gravity Forms confirmation
- Whoever got the url will go on confirmation, no authentication performed
- The confirmation will use latest field values from database which could have changed
- No payment status checking

### Redirect URL Retrieval Failure Handling

After form submit, this plugin sends order information to WorldPay in exchange for a redirect URL(the WorldPay hosted checkout form URL).

By default, when redirect URL retrieval fails:
1. Mark entry payment status as `Failed`
1. [Log](https://docs.gravityforms.com/logging-and-debugging/) the error
1. `wp_die` **immediately**

Common failure reasons:
- Incorrect vendor code
- Server IP not whitelisted

Tips: Check the [log](https://docs.gravityforms.com/logging-and-debugging/).


You can use `'gf_worldpay_redirect_url_failure_wp_die'` filter to:
- continue Gravity Forms' feed and confirmation flow
- perform extra operations
- redirect to a different error page

**Important:** If this filter returns `false`, normal Gravity Forms' feed and confirmation flow continues.
Improper settings might lead to disasters.

Example:
```php
add_filter('gf_worldpay_redirect_url_failure_wp_die', function(bool $shouldWpDie, ServerAuthorizeResponse $response, Entry $entry, GFPaymentAddOn $addOn): bool {

    // Do something.

    return true; // Do `wp_die`
    return false; // Don't `wp_die`, continue normal flow
    return $shouldWpDie; // Undecisive
}, 10, 4);
```

## Preflight

This plugin provides built-in support for [preflight-command](https://github.com/itinerisltd/preflight-command).
No extra setup steps required.


Checker ID: `gf-worldpay-production-mode`
- ensure all gf-worldpay feeds are in production mode (i.e: not in test mode)
- can't be disabled
- no config options available

## Coding

### Required Reading List

Read the followings before developing:

- [WorldPay HTML API](https://www.worldpay.com/uk/support/guides/business-gateway)
- [Gravity Forms: GFPaymentAddOn](https://docs.gravityforms.com/gfpaymentaddon/)
- [Gravity Forms: Entry Object](https://docs.gravityforms.com/entry-object/)
- [Omnipay: WorldPay](https://github.com/thephpleague/omnipay-worldpay)
- [thephpleague/omnipay#255 (comment)](https://github.com/thephpleague/omnipay/issues/255#issuecomment-90509446)

### Gravity Forms

Gravity Forms has undocumented hidden magics, read its source code.

## Author Information

[gf-worldpay](https://github.com/ItinerisLtd/gf-worldpay) is a [Itineris Limited](https://www.itineris.co.uk/) project created by [Tang Rufus](https://typist.tech).

Full list of contributors can be found [here](https://github.com/ItinerisLtd/gf-worldpay/graphs/contributors).

## Feedback

**Please provide feedback!** We want to make this library useful in as many projects as possible.
Please submit an [issue](https://github.com/ItinerisLtd/gf-worldpay/issues/new) and point out what you do and don't like, or fork the project and make suggestions.
**No issue is too small.**

## Change log

Please see [CHANGELOG](./CHANGELOG.md) for more information on what has changed recently.

## License

[gf-worldpay](https://github.com/ItinerisLtd/gf-worldpay) is released under the [MIT License](https://opensource.org/licenses/MIT).


CHANGELOG.md
# Change Log

## [0.2.5](https://github.com/ItinerisLtd/gf-worldpay/tree/0.2.5) (2019-02-18)
[Full Changelog](https://github.com/ItinerisLtd/gf-worldpay/compare/0.2.4...0.2.5)

**Merged pull requests:**

- Fix composer deprecation warning about uppercase characters [\#17](https://github.com/ItinerisLtd/gf-worldpay/pull/17) ([TangRufus](https://github.com/TangRufus))
- Readme: Update WorldPay setup instruction [\#16](https://github.com/ItinerisLtd/gf-worldpay/pull/16) ([TangRufus](https://github.com/TangRufus))

## [0.2.4](https://github.com/ItinerisLtd/gf-worldpay/tree/0.2.4) (2018-10-31)
[Full Changelog](https://github.com/ItinerisLtd/gf-worldpay/compare/0.2.3...0.2.4)

**Merged pull requests:**

- Version bump 0.2.4 [\#15](https://github.com/ItinerisLtd/gf-worldpay/pull/15) ([TangRufus](https://github.com/TangRufus))
- Better instructions and help text [\#14](https://github.com/ItinerisLtd/gf-worldpay/pull/14) ([TangRufus](https://github.com/TangRufus))
- Remove `CompletePurchaseRequest` & `CompletePurchaseResponse` logging [\#13](https://github.com/ItinerisLtd/gf-worldpay/pull/13) ([TangRufus](https://github.com/TangRufus))

## [0.2.3](https://github.com/ItinerisLtd/gf-worldpay/tree/0.2.3) (2018-09-29)
[Full Changelog](https://github.com/ItinerisLtd/gf-worldpay/compare/0.2.2...0.2.3)

**Merged pull requests:**

- Version bump 0.2.3 [\#12](https://github.com/ItinerisLtd/gf-worldpay/pull/12) ([TangRufus](https://github.com/TangRufus))
- Use meta type redirection instead of location header [\#11](https://github.com/ItinerisLtd/gf-worldpay/pull/11) ([TangRufus](https://github.com/TangRufus))

## [0.2.2](https://github.com/ItinerisLtd/gf-worldpay/tree/0.2.2) (2018-09-09)
[Full Changelog](https://github.com/ItinerisLtd/gf-worldpay/compare/0.2.1...0.2.2)

**Closed issues:**

-  Use new packagist.org auto-update mechanism [\#5](https://github.com/ItinerisLtd/gf-worldpay/issues/5)

**Merged pull requests:**

- Version bump 0.2.2 [\#9](https://github.com/ItinerisLtd/gf-worldpay/pull/9) ([TangRufus](https://github.com/TangRufus))
- Apply code style; Fix some minor bugs along the way [\#8](https://github.com/ItinerisLtd/gf-worldpay/pull/8) ([TangRufus](https://github.com/TangRufus))

## [0.2.1](https://github.com/ItinerisLtd/gf-worldpay/tree/0.2.1) (2018-08-22)
[Full Changelog](https://github.com/ItinerisLtd/gf-worldpay/compare/0.2.0...0.2.1)

**Merged pull requests:**

- Version bump 0.2.1 [\#7](https://github.com/ItinerisLtd/gf-worldpay/pull/7) ([TangRufus](https://github.com/TangRufus))
- :bike:Bikeshedding for open sourcing [\#6](https://github.com/ItinerisLtd/gf-worldpay/pull/6) ([TangRufus](https://github.com/TangRufus))

## [0.2.0](https://github.com/ItinerisLtd/gf-worldpay/tree/0.2.0) (2018-08-18)
[Full Changelog](https://github.com/ItinerisLtd/gf-worldpay/compare/0.1.0...0.2.0)

**Merged pull requests:**

- Version bump 0.2.0 [\#4](https://github.com/ItinerisLtd/gf-worldpay/pull/4) ([TangRufus](https://github.com/TangRufus))
- Add preflight checker; Remove build form source; Fix incorrect fallback to test mode [\#3](https://github.com/ItinerisLtd/gf-worldpay/pull/3) ([TangRufus](https://github.com/TangRufus))
- Update readme [\#2](https://github.com/ItinerisLtd/gf-worldpay/pull/2) ([TangRufus](https://github.com/TangRufus))

## [0.1.0](https://github.com/ItinerisLtd/gf-worldpay/tree/0.1.0) (2018-08-11)
**Merged pull requests:**

- Version bump 0.1.0 [\#1](https://github.com/ItinerisLtd/gf-worldpay/pull/1) ([TangRufus](https://github.com/TangRufus))



\* *This Change Log was automatically generated by [github_changelog_generator](https://github.com/skywinder/Github-Changelog-Generator)*

src/FeedSettingsFields.php
<?php

declare(strict_types=1);

namespace Itineris\WorldPay;

use GFFeedAddOn;

class FeedSettingsFields
{
    public static function toArray(GFFeedAddOn $addOn): array
    {
        return [
            [
                'fields' => [
                    [
                        'name' => 'feedName',
                        'label' => esc_html__('Name', 'gf-worldpay'),
                        'type' => 'text',
                        'class' => 'medium',
                        'required' => true,
                        'tooltip' => '<h6>' .
                            esc_html__('Name', 'gf-worldpay') .
                            '</h6>' .
                            esc_html__('Enter a feed name to uniquely identify this setup.', 'gf-worldpay'),
                    ],
                ],
            ],
            [
                'title' => esc_html__('WorldPay Settings', 'gf-worldpay'),
                'fields' => self::worldPaySettingsFields($addOn),
            ],
            [
                'title' => esc_html__('Products &amp; Services Settings', 'gf-worldpay'),
                'fields' => [
                    [
                        'name' => 'paymentAmount',
                        'label' => esc_html__('Payment Amount', 'gf-worldpay'),
                        'type' => 'select',
                        'choices' => $addOn->product_amount_choices(),
                        'required' => true,
                        'default_value' => 'form_total',
                        'tooltip' => '<h6>' .
                            esc_html__('Payment Amount', 'gf-worldpay') .
                            '</h6>' .
                            esc_html__(
                                "Select which field determines the payment amount, or select 'Form Total' to use the total of all pricing fields as the payment amount.",
                                'gf-worldpay'
                            ),
                    ],
                ],
            ],
            [
                'title' => esc_html__('Order Settings', 'gf-worldpay'),
                'fields' => self::orderSettingsFields(),
            ],
            [
                'title' => esc_html__('Other Settings', 'gf-worldpay'),
                'fields' => [
                    [
                        'name' => 'conditionalLogic',
                        'label' => esc_html__('Conditional Logic', 'gf-worldpay'),
                        'type' => 'feed_condition',
                        'tooltip' => '<h6>' .
                            esc_html__('Conditional Logic', 'gf-worldpay') .
                            '</h6>' .
                            esc_html__(
                                'When conditions are enabled, form submissions will only be sent to the payment gateway when the conditions are met. When disabled, all form submissions will be sent to the payment gateway.',
                                'gf-worldpay'
                            ),
                    ],
                ],
            ],
        ];
    }

    private static function worldPaySettingsFields(GFFeedAddOn $addOn): array
    {
        return [
            [
                'type' => 'select_custom',
                'name' => 'installationId',
                'label' => esc_html__('Installation ID', 'gf-worldpay'),
                'required' => true,
                'choices' => self::getAllChoices('installationId', $addOn),
                'tooltip' => esc_html__('The ID for this installation. For example, 204596.', 'gf-worldpay'),
            ],
            [
                'type' => 'select_custom',
                'name' => 'merchantCode',
                'label' => esc_html__('Merchant Code', 'gf-worldpay'),
                'required' => true,
                'choices' => self::getAllChoices('merchantCode', $addOn),
                'tooltip' => esc_html__(
                    'This specifies which merchant code should receive funds for this payment. Use random string if unsure.',
                    'gf-worldpay'
                ),
            ],
            [
                'type' => 'select_custom',
                'name' => 'cartId',
                'label' => esc_html__('Cart ID', 'gf-worldpay'),
                'required' => true,
                'choices' => self::getAllChoices('cartId', $addOn),
                'after_input' => esc_html__(
                    'Letters (A-Z and a-z) and Numbers(0-9); Maximum 255 characters',
                    'gf-worldpay'
                ),
                'tooltip' => esc_html__(
                    'Your own reference number for this purchase. It is returned to you along with the authorisation results by whatever method you have chosen for being informed (email and / or Payment Responses). Use form title if unsure.',
                    'gf-worldpay'
                ),
            ],
            [
                'type' => 'select_custom',
                'name' => 'md5Secret',
                'label' => esc_html__('MD5 Secret', 'gf-worldpay'),
                'required' => true,
                'choices' => self::getAllChoices('md5Secret', $addOn),
                'tooltip' => esc_html__(
                    '30-char MD5 secret for transactions field in the Integration Setup for your installation using the Merchant Interface > Installations option',
                    'gf-worldpay'
                ),
            ],
            [
                'type' => 'select_custom',
                'name' => 'paymentResponsePassword',
                'label' => esc_html__('Payment Response Password', 'gf-worldpay'),
                'required' => true,
                'choices' => self::getAllChoices('installationId', $addOn),
                'tooltip' => esc_html__(
                    'This 25-char password is used to validate a Payment Notifications message.',
                    'gf-worldpay'
                ),
            ],
            [
                'type' => 'text',
                'name' => 'description',
                'label' => esc_html__('Description', 'gf-worldpay'),
                'required' => true,
                'tooltip' => esc_html__('A brief description of the goods or services purchased.', 'gf-worldpay'),
            ],
            [
                'name' => 'isTest',
                'label' => esc_html__('Environment', 'gf-worldpay'),
                'required' => true,
                'type' => 'radio',
                'default_value' => 'test',
                'choices' => [
                    [
                        'label' => esc_html__('Production', 'gf-worldpay'),
                        'value' => 'production',
                    ],
                    [
                        'label' => esc_html__('Test', 'gf-worldpay'),
                        'value' => 'test',
                    ],
                ],
            ],
            [
                'type' => 'text',
                'name' => 'cancelUrl',
                'label' => esc_html__('Cancel URL', 'gf-worldpay'),
                'class' => 'large',
                'after_input' => esc_html__('Leave blank to use Gravity Forms confirmations', 'gf-worldpay'),
                'tooltip' => esc_html__(
                    'Enter the URL the user should be sent to if they cancelled the WorldPay checkout form or payment failed.',
                    'gf-worldpay'
                ),
            ],
        ];
    }

    private static function getAllChoices(string $key, GFFeedAddOn $addOn): array
    {
        $rawFeeds = $addOn->get_feeds();

        $feeds = array_map(function (array $rawFeed): Feed {
            return new Feed($rawFeed);
        }, $rawFeeds);

        $allChoices = array_map(function (Feed $feed) use ($key): string {
            return $feed->getSelectCustom($key);
        }, $feeds);

        $uniqueChoices = array_unique(
            array_filter($allChoices)
        );

        $choices = array_map(function (string $choice): array {
            return ['label' => $choice];
        }, $uniqueChoices);

        return array_merge([
            [
                'value' => '',
                'label' => 'Select a choice',
            ],
        ], $choices);
    }

    private static function orderSettingsFields(): array
    {
        return [
            [
                'name' => 'customerInformation',
                'label' => esc_html__('Customer Information', 'gf-worldpay'),
                'type' => 'field_map',
                'field_map' => self::customerInfoFields(),
                'tooltip' => '<h6>' .
                    esc_html__('Customer Information', 'gf-worldpay') .
                    '</h6>' .
                    esc_html__('Map your Form Fields to the available listed fields.', 'gf-worldpay'),
            ],
            [
                'name' => 'billingInformation',
                'label' => esc_html__('Billing Information', 'gf-worldpay'),
                'type' => 'field_map',
                'field_map' => self::addressFields(),
                'tooltip' => '<h6>' .
                    esc_html__('Billing Information', 'gf-worldpay') .
                    '</h6>' .
                    esc_html__('Map your Form Fields to the available listed fields.', 'gf-worldpay'),
            ],
        ];
    }

    private static function customerInfoFields(): array
    {
        return [
            [
                'name' => 'firstName',
                'label' => esc_html__('First Name', 'gf-worldpay'),
                'required' => true,
            ],
            [
                'name' => 'lastName',
                'label' => esc_html__('Last Name', 'gf-worldpay'),
                'required' => true,
            ],
            [
                'name' => 'email',
                'label' => esc_html__('Email', 'gf-worldpay'),
                'required' => false,
            ],
            [
                'name' => 'phone',
                'label' => esc_html__('Phone', 'gf-worldpay'),
                'required' => false,
            ],
        ];
    }

    private static function addressFields(): array
    {
        return [
            [
                'name' => 'address',
                'label' => esc_html__('Address', 'gf-worldpay'),
                'required' => true,
            ],
            [
                'name' => 'address2',
                'label' => esc_html__('Address 2', 'gf-worldpay'),
                'required' => false,
            ],
            [
                'name' => 'city',
                'label' => esc_html__('City', 'gf-worldpay'),
                'required' => true,
            ],
            [
                'name' => 'zip',
                'label' => esc_html__('Zip', 'gf-worldpay'),
                'required' => true,
            ],
            [
                'name' => 'country',
                'label' => esc_html__('Country', 'gf-worldpay'),
                'required' => true,
            ],
            [
                'name' => 'state',
                'label' => esc_html__('State', 'gf-worldpay'),
                'required' => false,
            ],
        ];
    }
}
